### ISSUE TITLE
feat: multiple application configurations, single vouch proxy instance 

### ISSUE THREAD

**Issue by vasi-grigo** [vouch/vouch-proxy#572]
First off, thank you for this project; i can see it making my devops life much easier :)
Secondly, are you guys considering adding a mode in which the proxy would run with multiple SSO configurations?
To add a bit more context:
- system runs in GKE
- nginx-ingress is used as the front
- vouch proxy is used as the SSO enforcer
- there are multiple applications each with its own SSO particulars (client_id, secret, etc., etc.)

Currently, to support this scenario, one has to deploy multiple instances of vouch-proxy. One per configuration to be exact.
This is:
- suboptimal in terms of resource usage (this is exacerbated by things like autopilot which sets minimum CPU usage to `250m`; so say, 10 apps will require `2500m` idling most of the time)
- hard to maintain; basically each application needs to either create another deployment alongside or use a sidecar

Ideas:
- upstream header; nginx (or whichever upstream server) sets a X-Vouch-Use-Config header that corresponds to a local config file inside the container where vouch-proxy is running  
- $host header; new configuration option like `multihost=true` causes vouch-proxy to run in multihost mode which uses $host to lookup its corresponding config file; e.g. request for `app0.foo.bar` uses `/config/app0.foo.bar.yaml`

It seems like something relatively easy to add to the project and would greatly benefit plenty of users apart from me and make it a bit more "cloud-native"
cheers

**Comment by adamency** [comment]
Hi @vasi-grigo. Have you been able to find a solution for this ?

To explain my usecase: I have been using oauth2-proxy myself instead, but to my understanding it can only serve multiple upstreams if they are hosted on different paths, i.e. using two different domains app1.dn.com and app2.dn.com isn't allowed, and we have to use dn.com/app1 and dn.com/app2 which breaks all application which don't support being served on a subpath.

So I'm also looking for an auth proxy which can do this.

**Comment by bnfinet** [comment]
This comes up from time to time and I just end up not being in love with this idea.  Feels like too much of a foot gun.

VP is fairly low resource usage.  Is it a pain to deploy multiple instances in parallel?  Yes it is.  But at least you can be sure that your request was routed to the correct security context.

**Comment by adamency** [comment]
@bnfinet It's no problem if you just have two apps to protect. But the thing is I have multiple apps which don't implement authentication themselves, multiplied by all our tenants. Which amounts to a couple dozens. Deploying 40 instances of an auth proxy on the cluster is indeed a complicated design decision to justify, or even consider.

But thanks for your answer anyway.

**Comment by vasi-grigo** [comment]
>This comes up from time to time and I just end up not being in love with this idea. Feels like too much of a foot gun.

how is it a foot gun to have all requests go through a centralized gateway for auth? (instead of having a separate deployment for each application)?

> VP is fairly low resource usage. Is it a pain to deploy multiple instances in parallel? Yes it is. But at least you can be sure that your request was routed to the correct security context.

it's not really about VP being low resource usage; in autopilot-like environments there's a default minimum for a pod which makes the math `min resource * num of vp instances`; for an HA setup this feature would be `min resource * 2`
you can be pretty sure that the request is securely/properly routed as each application would be under a separate DNS

**Comment by bnfinet** [comment]
@vasi-grigo thanks for your interest in VP

It sounds like it may be worth it to you both financially and on the merits to explore forking the codebase and seeing if you can achieve a multi tenant objective.  If you're able to do that, I would be so curious to learn about the solution you came up with and see if it's feasible for wider deployment. 

But I should set the expectation that I am skeptical that such a change would keep VPs config approachable, clear and secure and that I'm not personally interested in taking VP in that direction and I'm not at all interested in supporting the requests that would invariably present themselves as issues in this repository.  I'm sure many would agree that VP could use better support already.

If you would like to explore this concept further in this thread, could you please explore at least a concept for a working config based upon RBAC or ACL that might perform what you like and make a clearer proposal.  I doubt it'll make it into VP any time soon but perhaps it would serve to move this conversation forward and offer the next human to propose it a more productive starting point.

But honestly I have other things that I'd prefer to do with VP long before multi-tenancy.  I'm sorry that's the case.

**Comment by adamency** [comment]
@bnfinet Wow.

- I appreciated that you, the maintainer, answered that quickly after my message even if you said you would not consider adding this functionality cause at least you gave me the information, so even though you closed the issue right away I wanted to make an answer to show respect of your time.
- I still showed my appreciation by thanking you at the end.
- I detailed my usecase to clarify the need behind this feature request (as an attempt for healthy debate on this): i.e. that this feature is not a about resource usage, but scalability and infrastructure management.

And what you deem the relevant thing to do is to hide all my comments unceremoniously like this ? I seriously don't understand this move and feel like I completely misjudged you before, as it seems respect is not something you abide to yourself.
